spec/models/spree/payment_spec.rb

```diff
@@ -124,5 +124,86 @@ module Spree
+
+      # Mimicing call from PayPalController#confirm in spree_paypal_express
+      def create_new_paypal_express_payment(order, payment_method)
+        order.payments.create!({
+          :source => Spree::PaypalExpressCheckout.create({
+            :token => "123",
+            :payer_id => "456"
+          }, :without_protection => true),
+          :amount => order.total,
+          :payment_method => payment_method
+        }, :without_protection => true)
+      end
+
+      before do
+        order.update_totals
+      end
+
+      context "for paypal express payments with transaction fees" do
+        let!(:payment_method) { Spree::Gateway::PayPalExpress.create!(name: "PayPalExpress", distributor_ids: [create(:distributor_enterprise).id], environment: Rails.env) }
+        let(:calculator) { Spree::Calculator::FlatPercentItemTotal.new(preferred_flat_percent: 10) }
+
+        before do
+          payment_method.calculator = calculator
+          payment_method.save!
+        end
+
+        context "when a payment already exists on the order" do
+          let!(:existing_payment) do
+            order.payments.create!(payment_method_id: payment_method.id, amount: order.total)
+          end
+
+          context "and the existing payment uses the same payment method" do
+            context "and the existing payment does not have a source" do
+              it "removes the existing payment (and associated fees)" do
+                create_new_paypal_express_payment(order, payment_method)
+
+                order.reload
+                expect(order.payments.count).to eq 1
+                expect(order.adjustments.payment_fee.count).to eq 1
+                expect(order.adjustments.payment_fee.first.amount).to eq 1.5
+              end
+            end
+
+            context "and the existing payment has a source" do
+              let!(:source) { Spree::PaypalExpressCheckout.create(token: "123") }
+
+              before do
+                existing_payment.source = source
+                existing_payment.save!
+              end
+
+              it "does not remove the existing payment" do
+                create_new_paypal_express_payment(order, payment_method)
+
+                order.reload
+                expect(order.payments.count).to eq 2
+              end
+            end
+          end
+
+          context "and the existing payment uses a different method" do
+            let!(:another_payment_method) { Spree::Gateway::PayPalExpress.create!(name: "PayPalExpress", distributor_ids: [create(:distributor_enterprise).id], environment: Rails.env) }
+
+            before do
+              existing_payment.update_attributes(payment_method_id: another_payment_method.id)
+            end
+
+            it "does not remove the existing payment" do
+              create_new_paypal_express_payment(order, payment_method)
+
+              order.reload
+              expect(order.payments.count).to eq 2
```

Comment:
IMO it's better to check the actual payment objects so we avoid false positives.